specs/Standards/Tokens/NonFungibleToken.md

```diff
@@ -0,0 +1,196 @@
+
+```TypeScript
+[...]// [Corgi Contract]
+corgiToken.grantAccess(corgiTokenId, escrowID);
+
+[...] // [Sausage Contract]
+sausageContract.grantAccess(sausageTokenId, escrowID);
+
+[...] // [Escrow Contract]
+async escrowContractCall() {
+  await Promise.all([
+    corgiToken.transferFrom(tokenId, aliceID, jerryID);
+    sausageContract.transferFrom(tokenId, jerryID, aliceID);
+  ]);
+}
+
+```
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+## Template for smart contract in AssemblyScript
+
+
+At time of writing, this standard is established with several constraints found in AssemblyScript. The first is that interfaces are not an implemented feature of AssemblyScript, and the second is that classes are not exported in the conversion from AssemblyScript to WASM. This means that the entire contract could be implemented as a class, which might be better for code organization, but it would be deceiving in function.
+
+```TypeScript
+  type TokenId = u64;
+
+  /******************/
+  /* CHANGE METHODS */
+  /******************/
+  
+  // Grant the access to the given `accountId` for the given `tokenId`.
+  // Requirements:
+  // * The caller of the function (`predecessor_id`) should have access to the token.
+  export function grantAccess(tokenId: TokenId, accountId: string): void;
+
+  // Revoke the access to the given `accountId` for the given `tokenId`.
+  // Requirements:
+  // * The caller of the function (`predecessor_id`) should have access to the token.
+  export function revokeAccess(tokenId: TokenId, accountId: string): void;
+
+  // Transfer the given `tokenId` from the given `accountId`.  Account `newAccountId` becomes the new owner.
+  // Requirements:
+  // * The caller of the function (`predecessor_id`) should have access to the token.
+  export function transferFrom(tokenId: TokenId, accountId: string, newAccountID: string): void;
+
+
+  // Transfer the given `tokenId` to the given `accountId`.  Account `accountId` becomes the new owner.
+  // The token unlocks (if it was locked) and all access is revoked except for the new owner.
+  // Requirements:
+  // * The caller of the function (`predecessor_id`) should have access to the token.
+  export function transfer(tokenId: TokenId, accountId: string): void;
+
+  /****************/
+  /* VIEW METHODS */
+  /****************/
+
+// Returns `true` or `false` based on caller of the function (`predecessor_id) having access to the token
+  export function checkAccess(tokenId: TokenID): boolean;
+
+  // Get an individual owner by given `tokenId`.
+  export function getTokenOwner(tokenId: TokenId): string;
+
+  // Allows for the retrieval of tokens from given `ownerId` starting from given `startTokenIdx`.
+  // up to the provided `limit.`
+  // Requirements:
+  // * Starting index should not be larger than total number of tokens.
+  export function getTokensByOwner(ownerId: string, startTokenIdx: u64 = 0, limit: u32 = 10): Token[];
```

Comment:
We can't get this in the default implementation, because it needs complicated data structure to iterate on every token. Plus the initial owner need to own all tokens, which is not trivial as well.

So this can go into an extension.